halo2_proofs/src/plonk/circuit.rs

```diff
@@ -1545,6 +1685,472 @@ impl<F: Field> Gate<F> {
     }
 }
 
+/// Data that needs to be preprocessed from a circuit
+#[derive(Debug, Clone)]
+pub struct PreprocessingV2<F: Field> {
+    // TODO(Edu): Can we replace this by a simpler structure?
+    pub(crate) permutation: permutation::keygen::Assembly,
+    // TODO(Edu): Replace this by Vec<Vec<F>>.  Requires some methods of Polynomial to take Vec<F>
+    // instead
+    // pub(crate) fixed: Vec<Polynomial<F, LagrangeCoeff>>,
+    pub(crate) fixed: Vec<Vec<F>>,
+}
+
+/// This is a description of a low level Plonkish compiled circuit.  Contains the Constraint System
+/// as well as the fixed columns and copy constraints information.
+#[derive(Debug, Clone)]
+pub struct CompiledCircuitV2<F: Field> {
+    pub(crate) preprocessing: PreprocessingV2<F>,
+    pub(crate) cs: ConstraintSystemV2Backend<F>,
+}
+
+struct QueriesMap {
+    advice_map: HashMap<(Column<Advice>, Rotation), usize>,
+    instance_map: HashMap<(Column<Instance>, Rotation), usize>,
+    fixed_map: HashMap<(Column<Fixed>, Rotation), usize>,
+    advice: Vec<(Column<Advice>, Rotation)>,
+    instance: Vec<(Column<Instance>, Rotation)>,
+    fixed: Vec<(Column<Fixed>, Rotation)>,
+}
```

Comment:
This is probably nonsense. But It came to my mind the idea of having this struct as a `QueryBuilder` which returns a closure which given a reference to the `Queries` struct, can perform any search.

Unsure if this would be of any help or simplify anything. But just leaving it here.

Reply:
I think I don't fully understand your proposal, or at least, what I imagine from your description would be more complicated than the current code.

Reply:
Ohh more complicated for sure! But I wondered if this could be more performant/lightwheight.

In any case, happy to discuss offline!